### PR TITLE
make EntityHashMap and EntityHashSet proper types

### PR DESCRIPTION
**Pull request by Victoronz**
# Objective

`EntityHashMap` and `EntityHashSet` iterators do not implement `EntitySetIterator`.

## Solution

Make them newtypes instead of aliases. The methods that create the iterators can then produce their own newtypes that carry the `Hasher` generic and implement `EntitySetIterator`. Functionality remains the same otherwise.
There are some other small benefits, f.e. the removal of `with_hasher` associated functions, and the ability to implement more traits ourselves.

`MainEntityHashMap` and `MainEntityHashSet` are currently left as the previous type aliases, because supporting general `TrustedEntityBorrow` hashing is more complex. However, it can also be done.

## Testing

Pre-existing `EntityHashMap` tests.

## Migration Guide

Users of `with_hasher` and `with_capacity_and_hasher` on `EntityHashMap`/`Set` must now use `new` and `with_capacity` respectively.
If the non-newtyped versions are required, they can be obtained via `Deref`, `DerefMut` or `into_inner` calls.

### CODE REVIEW

File: crates/bevy_ecs/src/entity/hash_set.rs [hunk 1]
```diff
@@ -0,0 +1,364 @@
...
+
+    /// Returns the inner [`HashSet`].
+    pub fn into_inner(self) -> HashSet<Entity, EntityHash> {
+        self.0
+    }
+
+    /// Clears the set, returning all elements in an iterator.
+    ///
+    /// Equivalent to [`HashSet::drain`].
+    pub fn drain(&mut self) -> Drain<'_> {
```
Comment (chescock) [inline crates/bevy_ecs/src/entity/hash_set.rs:49]: Do you also want a wrapper for `ExtractIf`?

File: crates/bevy_ecs/src/entity/hash_map.rs [hunk 2]
```diff
@@ -0,0 +1,287 @@
...
+impl<V> IntoIterator for EntityHashMap<V> {
+    type Item = (Entity, V);
+    type IntoIter = hash_map::IntoIter<Entity, V>;
+
+    fn into_iter(self) -> Self::IntoIter {
+        self.0.into_iter()
+    }
+}
+
+impl<V: PartialEq> PartialEq for EntityHashMap<V> {
```
[the remark this answers is not in the document]
Comment (Victoronz) [reply crates/bevy_ecs/src/entity/hash_map.rs:145]: They could!

File: crates/bevy_ecs/src/entity/hash_map.rs [hunk 3]
```diff
@@ -0,0 +1,287 @@
...
+
+impl<V: Eq> Eq for EntityHashMap<V> {}
+
+/// An iterator over the keys of a [`EntityHashMap`] in arbitrary order.
+/// The iterator element type is `&'a Entity`.
+///
+/// /// This struct is created by the [`keys`] method on [`EntityHashMap`]. See its documentation for more.
+///
+/// [`keys`]: EntityHashMap::keys
+pub struct Keys<'a, V, S = EntityHash>(hash_map::Keys<'a, Entity, V>, PhantomData<S>);
```
[the remark this answers is not in the document]
Comment (Victoronz) [reply crates/bevy_ecs/src/entity/hash_map.rs:159]: It is! This `S` generic parameter is currently only here to express the purpose of these newtypes to the readers:
Technically a newtype alone is enough, but then the type will not communicate why it exists.
This is especially the case for the `EntitySetIterator` impls.

File: crates/bevy_ecs/src/entity/hash_set.rs (hunk 1 above)
Comment (Victoronz) [reply crates/bevy_ecs/src/entity/hash_set.rs:49]: yes!

File: crates/bevy_ecs/src/entity/hash_map.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (Victoronz) [reply crates/bevy_ecs/src/entity/hash_map.rs:159]: In the future we might also want to support more hashers, as we currently have multiple "trusted" `Hasher`s in the engine. However I think I'll leave that for when we support hashing more `TrustedEntityBorrow` types.
(That allows for switching `MainEntityHashMap`/`Set` as well)

**Review by chescock — APPROVED** [review]

**Review by alice-i-cecile — APPROVED** [review]
Annoying to need more code for this, but this is well-executed and I agree with the justification.

**Comment by Victoronz** [comment]
Yeah, all this code and the upcoming `EntityIndexMap`/`Set` versions could be removed if the hashbrown iterators carried this generic, which for them would probably be a small non-breaking change to a few dozen lines at most.

**Pull request merged**